cxx/util_distribution_variant.hh

```diff
@@ -32,15 +33,14 @@ enum class DistributionEnum {
   stringcat
 };
 
-enum class EmissionEnum { sometimes_bitflip, gaussian, simple_string };
-
 struct DistributionSpec {
   DistributionEnum distribution;
   std::map<std::string, std::string> distribution_args = {};
 };
 
 struct EmissionSpec {
-  EmissionEnum emission;
+  std::string emission_name;
```

Comment:
I think we should keep this an enum, consistent with DistributionSpec, so we can treat emission/distribution more consistently (which becomes convenient once NoisyRelation is integrated).

Reply:
I'd rather get rid of DistributionEnum than make this an enum.  Having them be enum's just means that the logic for getting a distribution or emission becomes spread out over two functions instead of concentrated in a single get_ function, and we have another place to modify (the list of enum values) when we want to add a new one.

Reply:
I think if we have a field that can take one of a limited set of values, that should be an enum and not a string. If we keep `ObservationEnum` in #79 (which I introduced for reasons described in the PR description), then we're modifying the logic in two places anyway when we add a new distribution/emission (which seems to me like not much additional burden). I would rather keep this as an enum and continue the discussion on #79, and then maybe replace the distribution/emission enums with strings simultaneously.

Reply:
+Srinivas for his vote on whether the first field of EmissionSpec should be an enum or string.

Reply:
Reasons to use enum here, to represent one of a limited set of values:
- Enum comparison is faster than string comparison when we're using this to build a cluster prior of the right type.
- Strings seem more bug prone, where an invalid value could be introduced somewhere along the line and caught only when the if/else bottoms out (this issue doesn't exist for enums).
- The compiler tells us if we failed to account for an enum value in a switch/case statement, unlike for strings.
- The (minimal IMO) extra complexity of using enums instead of strings is hidden from the user.
- If we keep `ObservationEnum`, the gain in simplicity of using string instead of enum for distributions/emissions is erased (which is why I think we should discuss this on #79, since that PR better illustrates the rationale for using enum).

Reply:
I also prefer enum here (for the reasons Emily mentioned).

Reply:
My feeling is that using an enum here is a premature optimization.

We have two functions here: one that takes a string and returns a spec, and another that takes a spec and returns an Emission object.  We have the choice between exposing both functions to the user (= one of us programmers, the only users relevant to this discussion), or only exposing their composition.  By exposing both functions, we introduce a new data type and the user now has two call two functions instead of one.

It is true that the user can cache the result of the first function and gain some small, probably insignificant speed improvement.  But it is my feeling that whenever we can reduce code complexity at the cost of a little speed, we should happily do so.

Note that this way of viewing things make it clear that there is no net gain in terms of code bugy-ness.  Any bugs in the composition of the two functions would still be there in one of the two.  In particular, any bad string value would still cause problems when converting from the string to spec.

Regarding ObservationEnum, it appears that it is only there to support observation_string_to_value, and I think we can support that functionality in other ways.  For example, adding the following lines to Distribution in distributions/base.hh:

ObservationVariant from_string(const string&s) {
  SampleType t;
  stringstream ss(s);
  ss >> t;
  return t;
}

Reply:
`ObservationEnum` is also used at the top of `irm.cc` to build a `CleanRelation` with the right template parameter.

If you're confident that strings are better here, I'd ask that you let me merge #79 and then you send a follow-up PR that removes `DistributionSpec` and `EmissionSpec` (and `ObservationSpec` if you want). Considering the overall system, I don't see how to make this work nicely with strings (though I see your argument that it's nicer locally in `get_emission.hh`). Then we can look at both proposals more concretely.